core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -1520,46 +1551,106 @@ public long loadMetadata(AlluxioURI path, boolean recursive)
+          long fileId = loadDirectoryMetadata(path, options);
+          InodeDirectory inode = (InodeDirectory) mInodeTree.getInodeById(fileId);
+          if (!inode.isDirectChildrenLoaded() && options.isLoadDirectChildren()) {
+            String[] files = ufs.list(ufsUri.getPath());
+            LoadMetadataOptions loadMetadataOptions = LoadMetadataOptions.defaults();
+            loadMetadataOptions.setLoadDirectChildren(false).setCreateAncestors(false);
+            for (int i = 0; i < files.length; i++) {
+              loadMetadata(path.join(files[i]), loadMetadataOptions);
+            }
+            inode.isDirectChildrenLoaded();
+          }
+          return fileId;
+        }
       }
-      return loadDirectoryMetadata(path, recursive);
     } catch (IOException e) {
       LOG.error(ExceptionUtils.getStackTrace(e));
       throw e;
     }
   }
 
   /**
+   * Loads metadata for the file identified by the given path from UFS into Alluxio.
+   *
+   * @param path the path for which metadata should be loaded
+   * @param resolution the UFS resolution of path
+   * @param options the load metadata options
+   * @return the file id of the loaded file
+   * @throws BlockInfoException if an invalid block size is encountered
+   * @throws FileDoesNotExistException if there is no UFS path
+   * @throws InvalidPathException if invalid path is encountered
+   * @throws InvalidFileSizeException if invalid file size is encountered
+   * @throws FileAlreadyCompletedException if the file is already completed
+   * @throws IOException if an I/O error occurs
+   * @throws AccessControlException if permission checking fails
+   */
+  @GuardedBy("mInodeTree")
+  private long loadFileMetadata(AlluxioURI path, MountTable.Resolution resolution,
+      LoadMetadataOptions options)
+      throws IOException, BlockInfoException, FileDoesNotExistException, InvalidPathException,
+      AccessControlException, FileAlreadyCompletedException, InvalidFileSizeException {
+    if (mInodeTree.inodePathExists(path)) {
+      return mInodeTree.getInodeByPath(path).getId();
+    }
+    AlluxioURI ufsUri = resolution.getUri();
+    UnderFileSystem ufs = resolution.getUfs();
+
+    long ufsBlockSizeByte = ufs.getBlockSizeByte(ufsUri.toString());
+    long ufsLength = ufs.getFileSize(ufsUri.toString());
+    // Metadata loaded from UFS has no TTL set.
+    CreateFileOptions createFileOptions =
+        CreateFileOptions.defaults().setBlockSizeBytes(ufsBlockSizeByte)
+            .setRecursive(options.isCreateAncestors()).setMetadataLoad(true).setPersisted(true);
+    try {
+      long fileId = createFile(path, createFileOptions);
+      CompleteFileOptions completeOptions = CompleteFileOptions.defaults().setUfsLength(ufsLength);
+      completeFile(path, completeOptions);
+      return mInodeTree.getInodeById(fileId).getId();
+    } catch (FileAlreadyExistsException e) {
+      LOG.error("FileAlreadyExistsException seen unexpectedly.", e);
```

Comment:
what is the purpose of that? also, Throwables.propagate(e) seems to be deprecated: http://google.github.io/guava/releases/snapshot/api/docs/com/google/common/base/Throwables.html#propagate(java.lang.Throwable)